pkg/phlaredb/block_querier.go

```diff
@@ -905,6 +938,78 @@ func (b *singleBlockQuerier) SelectMatchingProfiles(ctx context.Context, params
 	return iter.NewMergeIterator(maxBlockProfile, false, iters...), nil
 }
 
+// Series selects the series labels from this block.
+//
+// Note: It will select ALL the labels in the block, not necessarily just the
+// subset in the time range SeriesRequest.Start to SeriesRequest.End.
+func (b *singleBlockQuerier) Series(ctx context.Context, params *ingestv1.SeriesRequest) ([]*typesv1.Labels, error) {
+	sp, ctx := opentracing.StartSpanFromContext(ctx, "Series Block")
+	defer sp.Finish()
+
+	err := b.Open(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	names, err := b.index.LabelNames(nil)
+	if err != nil {
+		return nil, err
+	}
+
+	selectors, err := parseSelectors(params.Matchers)
+	if err != nil {
+		return nil, err
+	}
+
+	labelNameMap := make(map[string]struct{}, len(params.LabelNames))
+	for _, name := range params.LabelNames {
+		labelNameMap[name] = struct{}{}
+	}
+
+	names = lo.Filter(names, func(name string, index int) bool {
+		_, ok := labelNameMap[name]
+		if !ok {
+			return false
+		}
+
+		if selectors.matchesAll() {
+			return true
+		}
+
+		for _, selector := range selectors {
+			for _, matcher := range selector {
```

Comment:
This doesn't seem right, you should be applying matcher for the right matching label name. I'd like to see a method that can be tested that only takes an IndexReader, matches, and returns all series.

I think you should be using `PostingsForMatchers` and `Series(id storage.SeriesRef, lbls *phlaremodel.Labels, chks *[]ChunkMeta)` see `(b *singleBlockQuerier) SelectMatchingProfiles` for some examples.

Again `matcher.Matches` is supposed to be run on values, not names.

Reply:
I think Cyril is right, we should get postings via `PostingsForMatchers` and iterate through the series. I haven't tested it but I'd expect something like:

```go
	// get all relevant series.
	for postings.Next() {
		if _, err = b.index.Series(postings.At(), &lbls, &chks); err != nil {
			return nil, err
		}
		if len(params.LabelNames) > 0 {
			if x := lbls.WithLabels(params.LabelNames...); len(x) > 0 {
				// TODO:
				//  - Not sure if we want to keep empty label sets
				//  - Filter out duplicates
				resp.LabelsSet = append(resp.LabelsSet, &typesv1.Labels{
					// TODO: Double check lbls' memory in not reused after Clone.
					Labels: x.Clone(),
				})
			}
			continue
		}
		resp.LabelsSet = append(resp.LabelsSet, &typesv1.Labels{
			Labels: lbls.Clone(),
		})
	}
```